Feature/buyer protection

prempratapsingh opened the pull request:
## Description
Buyer protection widget is now added to the library with,

1. Figma based UI/UX specs
2. Integration to backend API for fetching and displaying shop specific buyer protection details
3. Unit tests

Also refactored Trustbadge view and widget views for improved unit test coverage. Sonar cloud reports now show `91% code coverage` 🚀 

Xcode unit test coverage reports however shows 92.3% code coverage. I will see why there is a difference b/w XCode code coverage and Sonar cloud code coverage.

superus8r commented on Sources/Trustylib/Utilities/GradeCalculator.swift:
suggestion: this business logic could be part of domain layer or a model for readability.
This can avoid a large number of Util files in future by keeping the business logic of relevant parts together. 🙂

prempratapsingh replied:
The same rating to grade conversion logic has to be used for shop grade and product grade widgets, this is why this I added this GradeCalculator for reusability and improved unit tests.

superus8r commented on Sources/Trustylib/Models/CurrencyCode.swift:
```diff
@@ -0,0 +1,47 @@
+     */
+    var code: String {
+        switch self {
+        case .euro: return "EURO"
```
praise: it's amazing to see how swift and kotlin code are similar, it makes code bases more similar and consistent 👍

superus8r reviewed (changes requested):
The code itself looks good 🙂 there is only one thing: It seems like there is code coverage reported for the "Test files". 

chore: Please exclude the test files from calculated coverage report.

Everything else looks good, added some suggestions in the comments. 🙂

prempratapsingh commented:
> The code itself looks good 🙂 there is only one thing: It seems like there is code coverage reported for the "Test files".
> 
> chore: Please exclude the test files from calculated coverage report.
> 
> Everything else looks good, added some suggestions in the comments. 🙂

hey @superus8r what do you mean by `Test Files`?

XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests. It doesn't consider artifect files like mock JSON/Text, etc.

The increase in code coverage % report is because I have refactored the SwiftUI views, Models and ViewModels in a way that more code blocks could be covered during tests runs. I also added good amount of new tests 🙂

superus8r commented:
Hey @prempratapsingh thank you for your answer.

> what do you mean by Test Files? 

Test Files are files under `Tests` group.

> XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests.

Could you please tell which groups (e.g. `Sources`, `Tests`, `Example`, etc) have active test coverage? 🙂

prempratapsingh commented:
> Hey @prempratapsingh thank you for your answer.
> 
> > what do you mean by Test Files?
> 
> Test Files are files under `Tests` group.
> 
> > XCTest test results and code coverage are based on the actual Swift/Objective-C code which are executed unit tests.
> 
> Could you please tell which groups (e.g. `Sources`, `Tests`, `Example`, etc) have active test coverage? 🙂

@superus8r XCTest doesn't work on the concept of covering a group `Sources`, `Tests`, `Example`, etc.
Rather it uses a TestTarget named `TrustylibTests` where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code. 

So in essence `TrustylibTests` has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.

<img width="458" alt="Screenshot 2023-04-03 at 2 15 32 PM" src="https://user-images.githubusercontent.com/27926337/229458774-53d8e468-dc2e-4c4a-8b34-1d4860c850e0.png">

superus8r commented:
@prempratapsingh 

> XCTest doesn't work on the concept of covering a group Sources, Tests, Example, etc.
Rather it uses a TestTarget named TrustylibTests where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code.

Thank you for clarifying that the `XCTest` uses `TrustylibTests` as TestTarget where unit test classes are added as pure Swift code, I understand that. 🙂✅

> So in essence TrustylibTests has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.

I also understand that the `TrustylibTests` only has Swift test cases and you agree that coverage reports must be based on the test cases. ✅
It's great to know you also agree with that, 🙂 however, along with that, the tests are also reporting coverage on themselves. This is extra and needs to be excluded. 

In a nutshell: tests should not report coverage on themselves and this should be excluded 🙂 
I also added a screenshot:
<img width="1177" alt="Screenshot 2023-04-03 at 11 01 48" src="https://user-images.githubusercontent.com/5621503/229464579-2c4e2e58-f6f3-4cf9-84bb-232cf5d8c2f7.png">

superus8r commented:
@prempratapsingh 
I approved this for now.
If it's okay, I can fix it myself later. 🙂

prempratapsingh commented:
> @prempratapsingh
> 
> > XCTest doesn't work on the concept of covering a group Sources, Tests, Example, etc.
> > Rather it uses a TestTarget named TrustylibTests where unit test classes (BuyerProtectionDetailsModelTests, BuyerProtectionViewModelTests, etc) are added as pure Swift code.
> 
> Thank you for clarifying that the `XCTest` uses `TrustylibTests` as TestTarget where unit test classes are added as pure Swift code, I understand that. 🙂✅
> 
> > So in essence TrustylibTests has only Swift test cases (shown in screenshot below) added to it and thus the test reports, code coverage reports are based on these test cases only.
> 
> I also understand that the `TrustylibTests` only has Swift test cases and you agree that coverage reports must be based on the test cases. ✅ It's great to know you also agree with that, 🙂 however, along with that, the tests are also reporting coverage on themselves. This is extra and needs to be excluded.
> 
> In a nutshell: tests should not report coverage on themselves and this should be excluded 🙂 @irockel could we please get a third opinion on this? I also added a screenshot: <img alt="Screenshot 2023-04-03 at 11 01 48" width="1177" src="https://user-images.githubusercontent.com/5621503/229464579-2c4e2e58-f6f3-4cf9-84bb-232cf5d8c2f7.png">

@superus8r the CircleCI configuration for groups to be covered (that you showed in attached screenshot) hasn't been changed. It is the same as when the iOS trustylib code coverage was 72%

The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference - https://github.com/trustedshops-public/etrusted-ios-trustbadge-library/commit/3b04f6dcbd6db35ca580913cd815cd15a9d0cd96

prempratapsingh commented:
> @prempratapsingh I approved this for now. If it's okay, I can fix it myself later. 🙂

@superus8r what fix your are referring to? Could you elaborate please.

superus8r commented:
@prempratapsingh 

> The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference

I understand and respect that 🙂 

This discussion is:
- about: `tests covering themselves` 
- not about: `increase in code coverage` ❌

Please do not mix these.

prempratapsingh commented:
> @prempratapsingh
> 
> > The increase in code coverage reports is primarily due to the refactoring I did last week for Views, Models, ViewModel and other related layers. Please see this code commit for reference
> 
> I understand and respect that 🙂
> 
> This discussion is:
> 
> * about: `tests covering themselves`
> * not about: `increase in code coverage` ❌
> 
> Please do not mix these.

@superus8r I am confused about why there is point/concern about `tests covering themselves` now?

Please see Xcode generated code coverage report (for `feature/buyer-protection` branch) in the in the screenshot below. This coverage report is purely by Xcode, there is no CodeCov or CircleCI envolved, still the coverage shows 93.6% :)

<img width="959" alt="Screenshot 2023-04-03 at 3 22 45 PM" src="https://user-images.githubusercontent.com/27926337/229477219-9048689d-7814-472c-8b27-8e5a86326cd5.png">

Anyways, if you feel like some CircleCI end of change is required for better test coverage, please go for it.

superus8r commented:
@prempratapsingh 
Thanks I will do that

hrayatnia commented:
In spirits of open source and contribution, two of the tests get failed with XCode 14.2 by just running tests and also by using Fastlane.
Iphone 11 simulator, Iphone 14 Pro simulator.

Test names: 

`testBuyerProtectionDetailsReturnsCorrectFormattedStringForProtectionAmount` on both devices, regardless of the approach

`testNetworkDataServiceAddsCorrectURLParameters` acted randomly but failed a couple of times.

<img width="258" alt="Screenshot 2023-04-03 at 12 01 38" src="https://user-images.githubusercontent.com/5239218/229482777-0641c913-2a79-4f7e-831b-79d022d6ffb0.png">

<img width="255" alt="PNG image" src="https://user-images.githubusercontent.com/5239218/229482821-72a81e74-c2ab-4ae5-ac1d-ba555a3e1cdc.png">

<img width="1247" alt="PNG image" src="https://user-images.githubusercontent.com/5239218/229482896-667f8ed2-6c1c-4ffd-b4a9-678a5eaf6d1f.png">



Also, to contribute to the discussion, the XCode test coverage will also be 91.8% if you turn on the coverage target.

<img width="1792" alt="Screenshot 2023-04-03 at 12 24 31" src="https://user-images.githubusercontent.com/5239218/229483493-f6ecdb9a-c98a-41da-8817-4c0ae2dd7f2a.png">

<img width="1792" alt="Screenshot 2023-04-03 at 12 24 42" src="https://user-images.githubusercontent.com/5239218/229483568-ed6f3e7d-4d44-4dd8-8304-4278f8c601a4.png">

superus8r commented:
@prempratapsingh fyi
